server/packetio.go

```diff
@@ -64,39 +64,60 @@ func newPacketIO(conn net.Conn) *packetIO {
 	return p
 }
 
-func (p *packetIO) readPacket() ([]byte, error) {
+func (p *packetIO) readOnePacket() ([]byte, error) {
 	var header [4]byte
 
 	if _, err := io.ReadFull(p.rb, header[:]); err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	length := int(uint32(header[0]) | uint32(header[1])<<8 | uint32(header[2])<<16)
-	if length < 1 {
-		return nil, errInvalidPayloadLen.Gen("invalid payload length %d", length)
-	}
-
 	sequence := uint8(header[3])
 	if sequence != p.sequence {
 		return nil, errInvalidSequence.Gen("invalid sequence %d != %d", sequence, p.sequence)
 	}
 
 	p.sequence++
 
+	length := int(uint32(header[0]) | uint32(header[1])<<8 | uint32(header[2])<<16)
+
 	data := make([]byte, length)
 	if _, err := io.ReadFull(p.rb, data); err != nil {
 		return nil, errors.Trace(err)
 	}
+	return data, nil
+}
+
+func (p *packetIO) readPacket() ([]byte, error) {
+	data, err := p.readOnePacket()
+	if err != nil {
+		return nil, err
+	}
+
+	length := len(data)
+
+	if length == 0 {
```

Comment:
@shenli 
If that's the case, `readPacket()` may just __ignore__ the zero-length check. 
It would be up to the caller of `readPacket()` to deal with empty packet (either as EOF or invalid).  
I will update with a test case( @c4pt0r )